test/gsb-test-srv/main.go

```diff
@@ -0,0 +1,468 @@
+	if !strings.HasSuffix(urlData.Url, "/") {
+		urlData.Url += "/"
+	}
+
+	newHp := newListEntry(urlData.Url)
+	match := t.list.findByHash(newHp.hash)
+	if match != nil {
+		log.Printf("Skipped adding %q - Already on the list\n", newHp.url)
+		return
+	}
+
+	// Lock the mutex for writing
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Add the new hash prefix and resort the list
+	t.list = append(t.list, newHp)
+	t.list.sort()
+	log.Printf("Added %q to safe browsing list\n", urlData.Url)
+}
+
+// getHits returns a JSON object describing how many times each URL on the mock
+// safebrowsing list was asked about by a client. E.g.
+// ```
+// {
+//   "evil.com": 2,
+//   "example.com": 0
+// }
+// ```
+func (t *testSrv) getHits(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Lock the mutex for reading
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	w.Header().Set("Content-Type", "application/json")
+	body, err := json.Marshal(t.hits)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if _, err := w.Write(body); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	log.Printf("Processed /hits request for client\n")
+}
+
+// gsbHandler creates an http.HandlerFunc that wraps the `inner` handler. The
+// wrapper will check that the request is a POST, has the correct content-type,
+// and that the `key` GET parameter matches the server's API key.
+func (t *testSrv) gsbHandler(inner http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// The HTTP method must be POST
+		if r.Method != "POST" {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+		// The Content-Type must be the protocol buffers mime type
+		if r.Header.Get("Content-Type") != protoMime {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		// There must be a "key" URL parameter with the correct API key
+		// TODO(@cpu): Send back a protocol-correct bad auth response
+		key := r.URL.Query().Get("key")
+		if key != t.apiKey {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+		inner(w, r)
+	})
+}
+
+// start() sets up the HTTP muxer and spawns a Go routine to ListenAndServe
+func (t *testSrv) start(listenAddr string) {
+	mux := http.NewServeMux()
+	mux.Handle("/v4/threatListUpdates:fetch", t.gsbHandler(t.threatListUpdateFetch))
+	mux.Handle("/v4/fullHashes:find", t.gsbHandler(t.fullHashesFind))
+	mux.Handle("/hits", http.HandlerFunc(t.getHits))
+	mux.Handle("/add", http.HandlerFunc(t.addSite))
+
+	go func() {
+		err := http.ListenAndServe(listenAddr, mux)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err.Error())
+			return
+		}
+	}()
+}
+
+// newTestServer constructs a testSrv instance with its list constructed from
+// the provided slice of strings.
+func newTestServer(apiKey string, unsafeURLs []string) testSrv {
+	ts := testSrv{
```

Comment:
I think it would be more typical to put this object at the bottom of the function so you can include `list:` as an inline value also.